Applications/Spire/source/charting/trend_line_model.cpp

```diff
@@ -0,0 +1,193 @@
+    }
+    return interval_start - threshold <= value && value <=
+      interval_end + threshold;
+  }
+
+  bool is_within_interval(Quantity value, Quantity interval_start,
+      Quantity interval_end) {
+    return is_within_interval(value, interval_start, interval_end,
+      Quantity(0));
+  }
+
+  Quantity y_intercept(Quantity x, Quantity y, Quantity slope) {
+    return y - (x * slope);
+  }
+
+  Quantity calculate_y(Quantity m, Quantity x, Quantity b) {
+    return (m * x) + b;
+  }
+
+  Quantity closest_point_distance_squared(Quantity x, Quantity y,
+      const TrendLine& line) {
+    auto pt1_distance = distance_squared(x, y,
+      static_cast<Quantity>(std::get<0>(line.m_points).m_x),
+      static_cast<Quantity>(std::get<0>(line.m_points).m_y));
+    auto pt2_distance = distance_squared(x, y,
+      static_cast<Quantity>(std::get<1>(line.m_points).m_x),
+      static_cast<Quantity>(std::get<1>(line.m_points).m_y));
+    return std::min(pt1_distance, pt2_distance);
+  }
+}
+
+TrendLineModel::TrendLineModel()
+  : m_last_id(0) {}
+
+int TrendLineModel::add(const TrendLine& line) {
+  m_trend_lines.emplace_back(TrendLineEntry{line,
+    ++m_last_id, State::UNSELECTED});
+  return m_last_id;
+}
+
+TrendLine TrendLineModel::get(int id) {
+  auto iter = find_id(id);
+  if(iter == m_trend_lines.end()) {
+    throw std::exception("TrendLine not found.");
+  }
+  return iter->m_trend_line;
+}
+
+std::vector<TrendLine> TrendLineModel::get_lines() const {
+  auto lines = std::vector<TrendLine>();
+  for(auto& line : m_trend_lines) {
+    lines.push_back(line.m_trend_line);
+  }
+  return lines;
+}
+
+std::vector<int> TrendLineModel::get_selected() const {
+  auto selected = std::vector<int>();
+  for(auto& line : m_trend_lines) {
+    if(line.m_state == State::SELECTED) {
+      selected.push_back(line.m_id);
+    }
+  }
+  return selected;
+}
+
+int TrendLineModel::intersects(const ChartPoint& point,
+    ChartValue threshold) const {
+  auto closest_id = -1;
+  auto closest_distance = std::numeric_limits<Quantity>::infinity();
+  auto threshold_squared = static_cast<Quantity>(threshold) *
+    static_cast<Quantity>(threshold);
+  auto point_x = static_cast<Quantity>(point.m_x);
+  auto point_y = static_cast<Quantity>(point.m_y);
+  for(auto& line : m_trend_lines) {
+    auto dist_squared = std::numeric_limits<Quantity>::infinity();
+    auto point_distance_squared = std::numeric_limits<Quantity>::infinity();
+    auto line_slope = slope(line.m_trend_line);
+    auto line_b = y_intercept(
+      static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_x),
+      static_cast<Quantity>(std::get<0>(line.m_trend_line.m_points).m_y),
+      line_slope);
+    point_distance_squared = closest_point_distance_squared(point_x,
+      point_y, line.m_trend_line);
+    if(std::isinf(static_cast<double>(line_slope))) {
+      if(is_within_interval(point_y,
+          static_cast<Quantity>(std::get<0>(
```

Comment:
Instead of constantly repeating ```static_cast<Quantity>(...)```, assign the expression to a variable and reuse that expression throughout.